nengo/builder/optimizer.py

```diff
@@ -0,0 +1,969 @@
+"""Operator graph optimizers."""
+
+from collections import defaultdict
+import logging
+import time
+import warnings
+
+import numpy as np
+
+from nengo.builder.neurons import SimNeurons
+from nengo.builder import operator
+from nengo.builder.operator import DotInc, ElementwiseInc, SlicedCopy
+from nengo.builder.signal import Signal
+from nengo.config import SupportRcDefaultsMixin
+from nengo.params import BoolParam, IntParam, RcDefault
+from nengo.utils.compat import zip_longest
+from nengo.utils.graphs import toposort, transitive_closure
+
+logger = logging.getLogger(__name__)
+
+
+class OpMergeOptimizer(SupportRcDefaultsMixin):
+    """Optimizes the operator graph by merging operators.
+
+    This reduces the number of iterators to iterate over in slow Python code
+    (as opposed to fast C code) and due to the merging operations will be
+    performed on larger chunks of sequential memory making better use of CPU
+    caching and prefetching.
+
+    The optimization algorithm has a worst case complexity of O(n^2 + e) where
+    n is the number of operators and e the number of edges in the dependency
+    graph. In practice the run time will be much better because not all n^2
+    pairwise combinations of operators will be evaluated. A grouping depending
+    on the operator type and view bases will be done with dictionaries. This
+    grouping can be done in amortized linear time and reduces the actual worst
+    case runtime of the optimization algorithm to O(g*m^2 + e) where g is the
+    number of groups and m the number of elements in a group. Moreover,
+    information about memory alignment will be used to cut the inner loop short
+    in many cases and gives a runtime much closer to linear in most cases.
+
+    Note that both `model` and `dg` will be changed by the `optimize` method.
+
+    Parameters
+    ----------
+    model : :class:`nengo.builder.builder.Model`
+        Builder output to optimize.
+    dg : dict
+        Dict of the form {a: {b, c}} where b and c depend on a specifying the
```

Comment:
Could we change this to "`{a: {b, c}}` where `b` and `c` depend on `a` specifying the" to make it easier to parse?